## Numpy Library Exercises

In [ ]:
import numpy as np

**1.** Given a random array, change the sign of elements whose values ​​are between 3 and 8

In [ ]:
arr = np.random.randint(10,size=(10))
print(arr)
arr[(arr>3) & (arr<8)] *= -1
print(arr)


[4 9 2 5 7 7 9 7 1 4]
[-4  9  2 -5 -7 -7  9 -7  1 -4]


**2.** Replace the maximum element of a random array with 0

In [ ]:
arr = np.random.randint(10,size=(10))
print(arr)
arr[arr==max(arr)] = 0
print(arr)

[7 3 6 1 5 7 0 9 7 4]
[7 3 6 1 5 7 0 0 7 4]


**3.** Construct the direct product of arrays (all combinations with each element). A two-dimensional array is provided as input.

In [ ]:
arr = np.random.randint(10,size=(3,2))
print(arr)
print('')
arrp = np.meshgrid(*arr, indexing='ij')
res = np.stack(arrp,axis=-1).reshape(-1,arr.shape[0])
print(res)


[[2 0]
 [3 4]
 [9 6]]

[[2 3 9]
 [2 3 6]
 [2 4 9]
 [2 4 6]
 [0 3 9]
 [0 3 6]
 [0 4 9]
 [0 4 6]]


**4.** Given two arrays, A (8x3) and B (2x2), find rows in A that contain elements from each row in B, regardless of the order of the elements in B.

In [ ]:
A = np.random.randint(10,size=(8,3))
B = np.random.randint(10,size=(2,2))
print(A)
print(B)
print("-" * 30)
for i, pom1 in enumerate(A):
  for j, pom2 in enumerate(B):
    if set(pom2).issubset(set(pom1)):
      print(f" Row {pom1} have elements of row {pom2}")

[[1 4 8]
 [8 3 7]
 [9 0 8]
 [8 7 3]
 [1 3 0]
 [0 8 1]
 [4 4 6]
 [5 1 8]]
[[5 1]
 [1 7]]
------------------------------
 Row [5 1 8] have elements of row [5 1]


**5.** Given a 10x3 matrix, find rows of unequal values ​​(for example, the row [2,2,3] remains, the row [3,3,3] is deleted)

In [ ]:
arr=np.random.randint(5, size=(10,3))
print(arr)
print("-" * 30)
res=[]
for i in range(len(arr)):
    if arr[i].min() != arr[i].max():
      res.append(arr[i])

print(np.array(res))

[[1 1 1]
 [3 1 0]
 [1 0 2]
 [1 1 0]
 [4 1 0]
 [3 1 0]
 [4 4 2]
 [1 2 4]
 [1 1 1]
 [3 2 4]]
------------------------------
[[3 1 0]
 [1 0 2]
 [1 1 0]
 [4 1 0]
 [3 1 0]
 [4 4 2]
 [1 2 4]
 [3 2 4]]


**6.** Given a two-dimensional array, remove the rows that are repeated.






In [ ]:
arr = np.random.randint(10,size=(10,3))
print(arr)
print('-'*30)
res = np.unique(arr,axis=0)
print(res)

[[6 1 1]
 [5 0 7]
 [3 2 5]
 [0 6 0]
 [1 7 2]
 [1 0 1]
 [7 7 6]
 [7 7 1]
 [2 4 5]
 [1 2 3]]
------------------------------
[[0 6 0]
 [1 0 1]
 [1 2 3]
 [1 7 2]
 [2 4 5]
 [3 2 5]
 [5 0 7]
 [6 1 1]
 [7 7 1]
 [7 7 6]]


______
______

For each of the following problems (1-5) you need to provide 2 implementations – one without using numpy (Consider that where the input or output should be a numpy array, there will simply be lists), and the second one is fully vectorized using numpy (without using Python loops/map/list comprehension).


__Note 1.__ It can be assumed that all specified objects are non-empty (for example, in __task 1__ there are non-zero elements on the diagonal of the matrix).

__Note 2.__ For most tasks, the solution takes no more than 1-2 lines.

___

* __Task 1__: Calculate the product of nonzero elements on the diagonal of a rectangular matrix.  
 For example, for X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]) answer is 3.

In [ ]:
#without numpy
X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]])
print(X)
shorter = min(len(X), len(X[0]))
prod = 1
for i in range(shorter):
  if X[i][i] != 0:
    prod*=X[i][i]

print("Product of nonzero elements: ",prod)

[[1 0 1]
 [2 0 2]
 [3 0 3]
 [4 4 4]]
Product of nonzero elements:  3


In [ ]:
#with numpy
X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]])
res = np.diag(X)
print(np.prod(res[res !=0]))

3


* __Task 2__: Two vectors x and y are given. Check if they define the same multiset.  
  For example, for x = np.array([1, 2, 2, 4]), y = np.array([4, 2, 1, 2]) answer is True.

In [ ]:
#withoutnp
x = np.array([1, 2, 2, 4])
y = np.array([4, 2, 1, 2])
if sorted(x) == sorted(y):
  print(True)
else:
  print(False)

True


In [ ]:
#withnp
x = np.array([1, 2, 2, 4])
y = np.array([4, 2, 1, 2])
res = np.array_equal(np.sort(x),np.sort(y))
print(res)

True


* __Task 3__: Find the maximum element in the vector x among the elements preceded by zero.
 For example, for x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0]) answer is 5.

In [ ]:
#withoutnp
x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0])
maxi = 0
for i in range(len(x)-1):
  if x[i] == 0:
    maxi = max(maxi,x[i+1])

print(maxi)


5


In [ ]:
#withnp
x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0])
res = np.max(x[1:][x[:-1]== 0])
print(res)

5


* __Task 4__: Implement run-length encoding. For some vector x, it is necessary to return a tuple of two vectors of the same length. The first one contains numbers and the second one contains how many times they need to be repeated.  
 For example, for x = np.array([2, 2, 2, 3, 3, 3, 5]) answer is (np.array([2, 3, 5]), np.array([3, 3, 1])).

In [ ]:
#withoutnp
x = np.array([2, 2, 2, 3, 3, 3, 5])
unix = []
dupx = []
cc=1
curr=x[0]
for i in range(1,len(x)):
  if x[i] == curr:
    cc+=1
  else:
    unix.append(curr)
    dupx.append(cc)
    curr = x[i]
    cc=1
unix.append(curr)
dupx.append(cc)
print(unix)
print(dupx)

[np.int64(2), np.int64(3), np.int64(5)]
[3, 3, 1]


In [ ]:
#withnp
x = np.array([2, 2, 2, 3, 3, 3, 5])
res = np.unique_counts(x)
print(res)

UniqueCountsResult(values=array([2, 3, 5]), counts=array([3, 3, 1]))


* __Task 5__: Two samples of objects are given - X and Y. Calculate the matrix of Euclidean distances between objects. Compare with function scipy.spatial.distance.cdist by operating speed.

In [ ]:
#withoutnp
import math
X=np.random.randint(20,size=(2,2))
Y=np.random.randint(20,size=(3,2))
res = []
print(X)
print(Y)
print("-"*30)
for i in X:
  row = []
  for j in Y:
    sqr_diff = sum((a-b) ** 2 for a, b in zip(i,j))
    distance = math.sqrt(sqr_diff)
    row.append(distance)
  res.append(row)
for i in res:
  print(i)


[[ 8  2]
 [ 1 18]]
[[10 18]
 [19  0]
 [ 8  0]]
------------------------------
[16.1245154965971, 11.180339887498949, 2.0]
[9.0, 25.45584412271571, 19.313207915827967]


_______
________

In [ ]:
#withnp
import time
from scipy.spatial.distance import cdist
X=np.random.rand(1000,2)
Y=np.random.rand(1000,2)
print(f"X: {X}")
print(f"Y: {Y}")
start = time.time()
distance = np.sqrt(np.sum((X[:,np.newaxis,:]-Y) ** 2,axis=2))
ntime = time.time() - start
print(f"Resuslt np: {distance}")
start = time.time()
distance_cdist = cdist(X,Y)
ctime = time.time() - start
print(f"Result scipy: {distance_cdist}")

print(f"Czas numpy: {ntime:.5f}")
print(f"Czas scipy: {ctime:.5f}")

X: [[0.33194572 0.42793472]
 [0.49330191 0.4866239 ]
 [0.15205917 0.04011863]
 ...
 [0.76184976 0.9722094 ]
 [0.20396767 0.33030718]
 [0.58050983 0.70116582]]
Y: [[0.01600941 0.12291006]
 [0.2358344  0.34299951]
 [0.12510859 0.60283815]
 ...
 [0.81058283 0.06997622]
 [0.21715763 0.96103333]
 [0.49404503 0.61909146]]
Resuslt np: [[0.4391535  0.12826292 0.27087415 ... 0.59768535 0.54531681 0.25063337]
 [0.6000799  0.29481771 0.38609853 ... 0.52370075 0.54892619 0.13246964]
 [0.15926067 0.31425327 0.56336454 ... 0.65920019 0.9232127  0.67243131]
 ...
 [1.13030403 0.82012026 0.73612121 ... 0.90354835 0.54480677 0.44318354]
 [0.27989618 0.03430137 0.28371092 ... 0.66011678 0.63086405 0.40931802]
 [0.80810918 0.49707571 0.4658955  ... 0.67181389 0.44671684 0.11921561]]
Result scipy: [[0.4391535  0.12826292 0.27087415 ... 0.59768535 0.54531681 0.25063337]
 [0.6000799  0.29481771 0.38609853 ... 0.52370075 0.54892619 0.13246964]
 [0.15926067 0.31425327 0.56336454 ... 0.65920019 0.9232127  0.672

* #### __Task 6__: CrunchieMunchies __*__

You work in the marketing department of a food company called MyCrunch, which is developing a new type of delicious, healthy cereal called **CrunchieMunchies**.

You want to show consumers how healthy your cereal is compared to other leading brands, so you've compiled nutrition data from several different competitors.

Your task is to use Numpy calculations to analyze this data and prove that your **СrunchieMunchies** - the healthiest choice for consumers.


In [ ]:
import numpy as np

1. View the cereal.csv file. This file contains calorie counts for various brands of cereal. Download the data from the file and save it as calorie_stats.

In [ ]:
calorie_stats = np.loadtxt("/cereal.csv", delimiter=",")
calorie_stats

array([ 70., 120.,  70.,  50., 110., 110., 110., 130.,  90.,  90., 120.,
       110., 120., 110., 110., 110., 100., 110., 110., 110., 100., 110.,
       100., 100., 110., 110., 100., 120., 120., 110., 100., 110., 100.,
       110., 120., 120., 110., 110., 110., 140., 110., 100., 110., 100.,
       150., 150., 160., 100., 120., 140.,  90., 130., 120., 100.,  50.,
        50., 100., 100., 120., 100.,  90., 110., 110.,  80.,  90.,  90.,
       110., 110.,  90., 110., 140., 100., 110., 110., 100., 100., 110.])

2. One serving of CrunchieMunchies contains 60 calories. How much higher is the average calorie count of your competitors?

Store the answer in the variable average_calories and print the variable in the terminal

In [ ]:
average_calories = np.average(calorie_stats)
print(f"Average calories of other companies is {average_calories}")
print(f"This is {round(average_calories-60,2)} higher than CrunchieMunchies")

Average calories of other companies is 106.88311688311688
This is 46.88 higher than CrunchieMunchies


3. Does the average calorie count accurately reflect the distribution of the data set? Let's sort the data and see.

Sort the data and store the result in the calorie_stats_sorted variable. Print the sorted information.

In [ ]:
sorted_calories = np.sort(calorie_stats)
print(f"Sorted calories: {sorted_calories}")

Sorted calories: [ 50.  50.  50.  70.  70.  80.  90.  90.  90.  90.  90.  90.  90. 100.
 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100.
 100. 100. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110.
 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110.
 110. 110. 110. 120. 120. 120. 120. 120. 120. 120. 120. 120. 120. 130.
 130. 140. 140. 140. 150. 150. 160.]


4. It looks like most values ​​are above the mean. Let's see if the median is the most accurate measure of the data set.

Calculate the median of the data set and save your answer in median_calories. Print the median so you can see how it compares to the mean.

In [ ]:
median_calories = np.median(calorie_stats)
print(f"Median value: {median_calories}")

Median value: 110.0


5. While the median shows that at least half of our values ​​are over 100 calories, it would be more impressive to show that a significant portion of the competitors have higher calorie counts than CrunchieMunchies.

Calculate various percentiles and print them until you find the lowest percentile that exceeds 60 calories. Store this value in the nth_percentile variable.

In [ ]:
for i in range(1,100):
  actual_p = np.percentile(calorie_stats,[i])
  print(f"Percentil {i} totals: {actual_p}")
  if actual_p > 60:
    nth_percentile = i
    print(f"The lowest percentile is: {nth_percentile}")
    break


Percentil 1 totals: [50.]
Percentil 2 totals: [50.]
Percentil 3 totals: [55.6]
Percentil 4 totals: [70.]
The lowest percentile is: 4


6. While the percentile shows us that most competitors have much higher calorie counts, it's an awkward concept to use in marketing materials.

Instead, let's calculate the percentage of cereals that contain more than 60 calories per serving. Save your answer in the more_calories variable and print it.

In [ ]:
more_calories = np.mean(calorie_stats > 60) * 100
print(f"The {more_calories:.0f}% companies cereals have more than 60 calories per serving")

The 96% companies cereals have more than 60 calories per serving


7. This is a really high percentage. This will be very useful as we promote CrunchieMunchies. But one question is: how big are the differences in the dataset? Can we generalize that most cereals contain around 100 calories or is the range even wider?

Calculate the magnitude of the deviation by finding the standard deviation. Save your answer in calorie_std and print it on the terminal. How can we incorporate this value into our analysis?

In [ ]:
calorie_std = np.std(calorie_stats)

print(f"Standard deviation totals: {calorie_std}")
print(f"Most of companies contain from {round(average_calories-calorie_std)} to {round(average_calories+calorie_std)} calories")

Standard deviation totals: 19.35718533390827
Most of companies contain from 88 to 126 calories


8. Please write a short paragraph outlining your findings and how you believe this data can be used to Mycrunch's advantage in marketing CrunchieMunchies.

CrunchieMunchies, with their 60 calories, are great, compared to the industry average of 106 calories. As many as 96% of competing cereals are higher in calories. Most other cereals have a calorie count between 87 and 125.

By leveraging these exact statistics in upcoming campaigns, Mycrunch can objectively position CrunchieMunchies not just as a slightly healthier alternative, but as an undeniable industry outlier.